File: backend/app/schemas/pipeline_run.py
"""Pydantic schemas for pipeline run API requests and responses.

Used for all three "PipelineRun-shaped" API responses: `GET
/api/pipeline/runs`, `GET /api/pipeline/runs/{id}`, and `POST
/api/pipeline/run` — see app/api/routes/pipeline.py for why the POST route
re-fetches the row instead of mapping `PipelineResult` directly.
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.enums import PipelineRunStatus


class PipelineRunRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    started_at: datetime
    finished_at: datetime | None
    status: PipelineRunStatus
    total_records: int
    valid_records: int
    invalid_records: int
    duplicate_records: int
    error_message: str | None
    input_file: str | None


class PipelineRunCreate(BaseModel):
    input_file: str | None = None


class PipelineInputsRead(BaseModel):
    files: list[str]
    default_file: str

File: frontend/src/app/core/models/pipeline-run.ts
/**
 * Mirrors the backend's `PipelineRunRead` response schema exactly
 * (backend/app/schemas/pipeline_run.py) — used for the response of
 * `POST /api/pipeline/run`, `GET /api/pipeline/runs`, and
 * `GET /api/pipeline/runs/{id}` alike, since the backend returns the same
 * shape from all three.
 */
export type PipelineRunStatus = 'queued' | 'started' | 'completed' | 'failed';

export interface PipelineRun {
  id: number;
  started_at: string;
  finished_at: string | null;
  status: PipelineRunStatus;
  total_records: number;
  valid_records: number;
  invalid_records: number;
  duplicate_records: number;
  error_message: string | null;
  input_file: string | null;
}
